src/ligo.ml

```diff
@@ -240,46 +240,94 @@ let geq_timestamp_timestamp = Z.geq
 
 (* tez *)
 
-type tez = Z.t
+type tez = Int64.t
 
 let tez_from_literal s =
   try
     let n = parse_int_with_suffix "mutez" s in
     if Z.lt n Z.zero then
-      failwith "Ligo.tez_from_literal: negative"
-    else n
+      failwith "Ligo.tez_from_literal: underflow"
+    else if Z.gt n (Z.of_int64 Int64.max_int) then
+      failwith "Ligo.tez_from_literal: overflow"
+    else
+      Z.to_int64 n
   with exc ->
     failwith ("Ligo.tez_from_literal: " ^ Printexc.to_string exc)
 
-let add_tez_tez = Z.add
+let add_tez_tez x y =
+  assert (x >= Int64.zero);
+  assert (y >= Int64.zero);
+  if x > Int64.sub Int64.max_int y then
+    failwith "Ligo.add_tez_tez: overflow"
+  else
+    Int64.add x y
 
 let sub_tez_tez x y =
-  if Z.lt x y then
-    failwith "Ligo.sub_tez_tez: negative"
+  assert (x >= Int64.zero);
+  assert (y >= Int64.zero);
+  if x < y then
+    failwith "Ligo.sub_tez_tez: underflow"
   else
-    Z.sub x y
-
-let mul_nat_tez = Z.mul
-
-let mul_tez_nat = Z.mul
+    Int64.sub x y
+
+let mul_nat_tez x y =
+  assert (x >= Z.zero);
+  assert (y >= Int64.zero);
+  if y = Int64.zero then
+    Int64.zero
+  else if x > Z.of_int64 (Int64.div Int64.max_int y) then
```

Comment:
Will this condition also work for cases where max_int cannot be divided evenly by y?

Reply:
After thinking a bit more about the semantics of Int64.div I think this might be okay as-is, but I still just wanted to double check.

Reply:
Yeah, since `Int64.div` floors and all numbers involved are non-negative, I believe this is the safest.